### DISCUSSION TITLE
Meaning of "supported diagnostics" and "fixable diagnostics" in CLI Tool

### DISCUSSION THREAD

**Discussion by olapiv** [JosefPihrt/Roslynator#776] [Q&A]
Hi!

I'm looking into this tool for the first time and generally it looks very cool! I was looking for a CLI tool to apply static analysis warnings and it looks like the Roslynator.CommandLine tool is exactly what I'm looking for.

However, I'm giving it a few test runs with the commands `analyze` vs `fix` with verbosity `diag` and there are a few details that I don't understand. I'm testing it with a minimal single-project solution, targeting .NET 5.0, and I'm running on Windows. I did tried this with a few analyzers from NuGet packages, but the following describes the results from using only the integrated `Microsoft.CodeAnalysis.CSharp.NetAnalyzers` (v 5.0.3.2303) and `Microsoft.CodeAnalysis.NetAnalyzers` (v 5.0.3.2303) assemblies.

What is confusing me, is essentially the output `x supported diagnostics with prefix y` and `xx fixable diagnostics with prefix yy`, which is shows in the console just before Roslynator compiles the project. The thing is that the number of `supported diagnostics` varies between `analyze` and `fix`.

Example:
```
Roslynator.exe analyze --msbuild-path [...] -v diag
// —> Supported: CA1067 Info, CA1401 Info, CA1016 Info
Roslynator.exe fix --msbuild-path [...] -v diag
// —> Supported: CA1067 Info, CA1401 Info
// —> Fixable: CA1067, CA1001
```
_This example only cherry-picks a few diagnostic IDs to make the point._

It seems like supported diagnostics in `fix` is always a subset of supported diagnostics in `analyze`. The set of fixable diagnostics sometimes adds diagnostic IDs, which are not mentioned in neither sets beforehand (CA1001 in this example), but also occasionally removes a diagnostic ID, which was previously supported (CA1401 in this example). I at first thought these three sets could be classified as 

1. Supported diagnostics (project has corresponding DiagnosticAnalyzer)
2. Fixable supported diagnostics (project has both DiagnosticAnalyzer and CodeFixProvider)
3. Fixable total diagnostics (project has CodeFixProvider, but not necessarily corresponding DiagnosticAnalyzer)

However, this would not explain that CA1401 is missing for "fixable".

Is there anything that I'm clearly misinterpreting here? Thanks! :)

**Comment by josefpihrt** [comment]
Greetings to Sweden,

First of all I'm sorry for a bit late answer.

> 1. Supported diagnostics (project has corresponding DiagnosticAnalyzer)
> 2. Fixable supported diagnostics (project has both DiagnosticAnalyzer and CodeFixProvider)
> 3. Fixable total diagnostics (project has CodeFixProvider, but not necessarily corresponding DiagnosticAnalyzer)

Your description seems to be correct.

It is correct that supported diagnostics for `fix` command does not include diagnostics that does not have CodeFixProvider:
https://github.com/JosefPihrt/Roslynator/blob/f953e766de0aca1bbb15932ade9f89ee76f9f6b6/src/Workspaces.Core/CodeFixes/CodeFixer.cs#L185-L187

If `supported-diagnostics` or `ignored-diagnostics` parameter is used then supported/fixable diagnostics are filtered according to this parameter: 
https://github.com/JosefPihrt/Roslynator/blob/f953e766de0aca1bbb15932ade9f89ee76f9f6b6/src/Workspaces.Core/CodeAnalysisHelpers.cs#L44-L49


Regarding CA1401: I've installed `Microsoft.CodeAnalysis.NetAnalyzers` and it seems that CA1401 does not have a fix.

**Reply by olapiv** [reply]
Hi, thanks for the reply! The code samples definitely helped.

However, I still don't entirely understand - if CA1401 does not have a fix, why is it shown in the 2nd set ("Fixable supported diagnostics")?

**Reply by josefpihrt** [reply]
Hi, you are right, it's a bug and the fix is already prepared.

Thanks for noticing that.

**Reply by olapiv** [reply]
Cool, thanks for taking the time! Could you maybe reference the PR here?

**Reply by josefpihrt** [reply]
https://github.com/JosefPihrt/Roslynator/commit/a24e168a8e38c8d6c3834cf2083b5e51e2f16342
